linera-core/src/client/mod.rs

```diff
@@ -1132,79 +1181,133 @@ where
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
+        let mut tracker = MultichainTracker::new(tracker);
+        let remote_log = info.requested_received_log.clone();
+        let remote_node_chains_view = info.requested_received_log.into_iter().fold(
+            BTreeMap::<ChainId, Vec<BlockHeight>>::new(),
+            |mut chain_to_info, entry| {
+                chain_to_info
+                    .entry(entry.chain_id)
+                    .or_default()
+                    .push(entry.height);
+                chain_to_info
+            },
+        );
+
+        // Collect all certificates that have been successfully processed.
         let mut certificates: Vec<Certificate> = Vec::new();
-        let mut new_tracker = tracker;
 
-        for entry in info.requested_received_log {
-            let query = ChainInfoQuery::new(entry.chain_id)
-                .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
+        for (chain_id, mut block_batch) in remote_node_chains_view {
+            block_batch.sort();
 
-            let local_response = self
-                .client
-                .local_node
-                .handle_chain_info_query(query.clone())
-                .await
-                .map_err(|error| NodeError::LocalNodeQuery {
-                    error: error.to_string(),
-                })?
-                .info;
-            if !local_response.requested_sent_certificate_hashes.is_empty() {
-                // We've already processed incoming messages for this certificate.
-                new_tracker += 1;
+            self.advance_with_local(chain_id, &mut block_batch, &mut tracker)
+                .await?;
+
+            let Some(first_block) = block_batch.first() else {
+                // `advance_with_local` might have drained the whole `block_batch`.
+                // In that case, move to the next chain batch.
                 continue;
-            }
+            };
+            let batch_size = block_batch.last().unwrap().saturating_sub(*first_block).0 + 1; // safe to unwrap because we checked that the vec is not empty.
```

Comment:
nit: We try to replace unwrap() + comment by expect(reason) for instance: `.expect("block_batch should not be empty")`

Reply:
Got it. Will do in the future.  

This particular line already disappears in the following PR that Andreas made.